Proposal: define the CRD of nested components (#11)
kubernetes-sigs/cluster-api-provider-nested: proposals/20201026-creating-control-plane.md

    - [Bootstrap](#bootstrap)
    - [Creation](#creation)
      - [Native way](#native-way)
      - [Using the third-party provider](#using-the-third-party-provider)
    - [Control Plane Custom Resources](#control-plane-custom-resources)
      - [NestedEtcd CRD](#nestedetcd-crd)
      - [NestedKAS CRD](#nestedkas-crd)
      - [NestedKCM CRD](#nestedkcm-crd)
    - [Security Model](#security-model)
  - [Alternatives](#alternatives)
  - [Implementation History](#implementation-history)

## Glossary

Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).

The followings are terms that will be used in the proposal. Whether including them into the Glossary needs further discussion:

* **Super Cluster** - The underlying cluster that manages the physical nodes, all pods created through the NestedControlPlanes will run on this cluster.
* **NestedControlPlane(NCP)** - The control plane that are hosted on the super cluster. 
* **NestedControlPlaneTemplate(NCPT)** - NCPT contains all information required to create the NCP.
* **NestedEtcd(NEtcd)** - The etcd that belongs to the control plane of the nested cluster.

> * **NestedKAS(NKAS)** - The kube-apiserver which belongs to the control plane of the nested cluster.

Author: Yes, I agree. having `nested` sound verbose. Will rename the CRD

> * **NestedKAS(NKAS)** - The kube-apiserver which belongs to the control plane of the nested cluster.

Reviewer: Are we going to support non-K8s `APIServer` and `ControllerManager` as well?
And are we going to folk and customize these components?

If not, having a `kube` prefix looks verbose too and may indicate a different message to users/developers as well

> * **NestedKAS(NKAS)** - The kube-apiserver which belongs to the control plane of the nested cluster.

Author: @brightzheng100 No, we only support the standard `apiserver` and `kcm`. Will change to `APIServer` and `ControllerManager`

> * **NestedKAS(NKAS)** - The kube-apiserver which belongs to the control plane of the nested cluster.

Reviewer 2: +1 to supporting only standard Kubernetes API Server and controller manager, that reduces the scope and keeps us in focus.

`Nested` prefix might be needed if we want to reuse the default API groups, e.g. `infrastructure.cluster.x-k8s.io` which RBAC is automatically setup for you.

Otherwise, if we want to use a different API group, and we drop the prefix, we need to setup RBAC in a way that CAPI manager can have permissions to act on these objects

> * **NestedKAS(NKAS)** - The kube-apiserver which belongs to the control plane of the nested cluster.

Reviewer 3: Thanks for that note @vincepri, that was my one concern. Nomenclature wise we can keep `Nested*` if that will cause issues.

> * **NestedKAS(NKAS)** - The kube-apiserver which belongs to the control plane of the nested cluster.

Reviewer 3: Agree @charleszheng44 @brightzheng100 `NestedAPIServer` `NestedControllerManager` and, `NestedEtcd`.

* **NestedKCM(NKCM)** - The kube-control-manager which belongs to the control plane of the nested cluster.
* **Sub-controller** - The controller that are incharge of NCP components, including NEtcd controller, NKAS controller and NKCM controller.
* **Component Provider** - Operators that creates components of the NCP.

## Summary

The goal of this proposal is to define CRDs of the three major components (kube-apiserver~(KAS), Etcd, kube-controller-manager~(KCM)) of the NCP, and a standard process of creating them.

## Motivation

> CAPN aims at providing control-plane level isolation while sharing physical resources among control planes. There exist various approaches to creating isolated control-planes. For example, one can run components of the nested control-plane as pods on the underlying clusters, create NCPs through cloud providers' Kubernetes services or use third-party component providers to create each component. In this proposal, we try to define CRDs of the NCP's three major components and a standard process of creating the three components regardless of which underlying approach is used. As examples, we introduce two setups that 1) creating each component natively, 2) creating NKAM and NKCM natively while using the [Etcd-cluster-operator](https://github.com/improbable-eng/etcd-cluster-operator) to create the NEtcd.

Reviewer 2: Consider also dropping the `-` between control plane

> CAPN aims at providing control-plane level isolation while sharing physical resources among control planes. There exist various approaches to creating isolated control-planes. For example, one can run components of the nested control-plane as pods on the underlying clusters, create NCPs through cloud providers' Kubernetes services or use third-party component providers to create each component. In this proposal, we try to define CRDs of the NCP's three major components and a standard process of creating the three components regardless of which underlying approach is used. As examples, we introduce two setups that 1) creating each component natively, 2) creating NKAM and NKCM natively while using the [Etcd-cluster-operator](https://github.com/improbable-eng/etcd-cluster-operator) to create the NEtcd.

Reviewer 3: Agree @vincepri

> CAPN aims at providing control-plane level isolation while sharing physical resources among control planes. There exist various approaches to creating isolated control-planes. For example, one can run components of the nested control-plane as pods on the underlying clusters, create NCPs through cloud providers' Kubernetes services or use third-party component providers to create each component. In this proposal, we try to define CRDs of the NCP's three major components and a standard process of creating the three components regardless of which underlying approach is used. As examples, we introduce two setups that 1) creating each component natively, 2) creating NKAM and NKCM natively while using the [Etcd-cluster-operator](https://github.com/improbable-eng/etcd-cluster-operator) to create the NEtcd.

Author: hyphen removed

### Goals

- Define the CRD that represents each nested component. The CRD needs to meet two requirements:

>     * Portable - the CRD should hold information that are required by different component providers, e.g., [etcd-operator](https://github.com/coreos/etcd-operator), [etcd-cluster-operator](https://github.com/improbable-eng/etcd-cluster-operator/blob/f84abc6561735814debd67d45bb62d2d2ed8cf4a/api/v1alpha1/etcdcluster_types.go#L31-L47)

Reviewer 2: We might want to add `etcdadm` as well here?

>     * Portable - the CRD should hold information that are required by different component providers, e.g., [etcd-operator](https://github.com/coreos/etcd-operator), [etcd-cluster-operator](https://github.com/improbable-eng/etcd-cluster-operator/blob/f84abc6561735814debd67d45bb62d2d2ed8cf4a/api/v1alpha1/etcdcluster_types.go#L31-L47)

Author: will add `etcdadm`

>     * Portable - the CRD should hold information that are required by different component providers, e.g., [etcd-operator](https://github.com/coreos/etcd-operator), [etcd-cluster-operator](https://github.com/improbable-eng/etcd-cluster-operator/blob/f84abc6561735814debd67d45bb62d2d2ed8cf4a/api/v1alpha1/etcdcluster_types.go#L31-L47)

Author: > We might want to add `etcdadm` as well here?

will add the `etcdadm`

>     * Portable - the CRD should hold information that are required by different component providers, e.g., [etcd-operator](https://github.com/coreos/etcd-operator), [etcd-cluster-operator](https://github.com/improbable-eng/etcd-cluster-operator/blob/f84abc6561735814debd67d45bb62d2d2ed8cf4a/api/v1alpha1/etcdcluster_types.go#L31-L47)

Reviewer 3: I want to be clear though that the goals are not to create these implementations, just not disallow someone else from doing this work out-of-tree.

>     * Portable - the CRD should hold information that are required by different component providers, e.g., [etcd-operator](https://github.com/coreos/etcd-operator), [etcd-cluster-operator](https://github.com/improbable-eng/etcd-cluster-operator/blob/f84abc6561735814debd67d45bb62d2d2ed8cf4a/api/v1alpha1/etcdcluster_types.go#L31-L47)

Reviewer 2: Let's add some text to that effect, it'd be good to clarify

    * Customizable - the CRD should allow end-users to customize each component, i.e., specify the image, component version, and command-line options.

- Define a standard process of creating nested components for NCP.

> 

Reviewer 3: ```suggestion

- Support independently creating/updating each component
```

- Present two examples of creating NCP natively or using the third-party provider to create the NEtcd.

### Non-Goals

- Define how NCP controller works.

- Discuss the details of how the third-party component provider work.

## Proposal

### Portability and Customizability

Generally, creating the three major components requires similar high-level information, like the components' version, the number of replicas, and the amount of computing resources. Meanwhile, end-users should be able to customize NCP components, i.e., specifying the component image, version, and command-line options. Therefore, we define a new struct `NestedComponentSpec` that contains common information required by different providers as well as customized information specified by the end-users. The `NestedComponentSpec` will look like the following

```go
type NestedComponentSpec struct {
    // ComponentSpec defines the common information for creating the component
    // +optional

>     CommonSpec addonv1alpha1.CommonSpec `json:"commonSpec",omitempty`

Reviewer 2: Is the CommonSpec going to be used anywhere else? If not, we might want to just inline the fields instead?

>     CommonSpec addonv1alpha1.CommonSpec `json:"commonSpec",omitempty`

Reviewer 3: Oh thanks, my updates were wrong. 🤦‍♂️

    
    // Resources defines the amount of computing resources that will be used by this component
    // +optional
    Resources corev1.ResourceRequirements `json:"resources",omitempty`
    
    // Replicas defines the number of replicas in the component's workload 
    // +optional
    Replicas int32 `json:"replicas",omitempty`

    // PatchSpecs includes the user specifed settings
    // +optional

>     PatchSpec addonv1alpha1.PatchSpec `json:"patchSpec,omitempty"`

Reviewer 2: I'd just inline this as well, instead of having a different struct

>     PatchSpec addonv1alpha1.PatchSpec `json:"patchSpec,omitempty"`

Reviewer 3: Oh thanks, my updates were wrong. 🤦‍♂️ (edited)

>     PatchSpec addonv1alpha1.PatchSpec `json:"patchSpec,omitempty"`

Reviewer 2: Oh, I actually meant to drop `PatchSpec` altogether if possible and just have the fields in here, it'd be easier to use

>     PatchSpec addonv1alpha1.PatchSpec `json:"patchSpec,omitempty"`

Reviewer 3: We can, although I'd like to think about actually supporting all the patching/mutation funcs from https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/ so that we don't have to immediately expose all the different controls for mutating the individual components.

For example my apiservers might need to turn on `DualStack` or another featuregate which might not be exposed as a typed field in the spec just yet, this allows us to be more flexible upfront and overtime "graduate" fields to the top-level component CRs. 

I have an item on the ClusterAddons agenda next week (https://docs.google.com/document/d/10_tl_SXcFGb-2109QpcFVrdrfnVEuQ05MBrXtasB0vk/edit#heading=h.8qktiog8x5ri) to bring up our use cases and gather any concerns they might have, do you think allowing this upfront concerning?

}
```

The `CommonSpecs` and the `PatchSpec` are defined in [kubebuilder-declarative-pattern](https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/1cbf859290cab81ae8e73fc5caebe792280175d1/pkg/patterns/addon/pkg/apis/v1alpha1/common_types.go):

```go
// CommonSpec defines the set of configuration attributes that must be exposed on all addons.
type CommonSpec struct {
	// Version specifies the exact addon version to be deployed, eg 1.2.3
	// It should not be specified if Channel is specified
	Version string `json:"version,omitempty"`
	// Channel specifies a channel that can be used to resolve a specific addon, eg: stable
	// It will be ignored if Version is specified
	Channel string `json:"channel,omitempty"`
}

// +k8s:deepcopy-gen=true
type PatchSpec struct {
	Patches []*runtime.RawExtension `json:"patches,omitempty"`
}
```

> If the designated component's providers do not allow patching, the `PatchSpec` will be ignored. All Nested CRs need to contain a `NestedComponentSpec`. We will not discuss how NCPT and NCP controller generates the `NestedComponentSpec` in this proposal. However, a possible way is to add three fields to the NCPT CRD, and has each of them holds the `NestedComponentSpec` for one of the components. The NCPT will look like the following

Author: I thought the Nested CRs would be created by NCP controller using NCPT. Will remove this section.

> If the designated component's providers do not allow patching, the `PatchSpec` will be ignored. All Nested CRs need to contain a `NestedComponentSpec`. We will not discuss how NCPT and NCP controller generates the `NestedComponentSpec` in this proposal. However, a possible way is to add three fields to the NCPT CRD, and has each of them holds the `NestedComponentSpec` for one of the components. The NCPT will look like the following

Reviewer 2: We can also use webhook validation to disallow certain fields to be filled, so that data doesn't get ignored silently and feedback is given to users immediately

> If the designated component's providers do not allow patching, the `PatchSpec` will be ignored. All Nested CRs need to contain a `NestedComponentSpec`. We will not discuss how NCPT and NCP controller generates the `NestedComponentSpec` in this proposal. However, a possible way is to add three fields to the NCPT CRD, and has each of them holds the `NestedComponentSpec` for one of the components. The NCPT will look like the following

Reviewer 3: Yeah, this would end up being up to the controller re-implementer to be responsible for as well which makes CAPNs responsibility here significantly lessened.

```go
// NestedControlPlaneTemplateSpec defines the desired state of NestedControlPlaneTemplate

> type NestedControlPlaneTemplateSpec struct {

Reviewer: `Template` may not be a good name here, not sure.

But I personally like the idea in VC where we have two different concepts: `ClusterVersions` CR and `VirtualCluster` CR.
This easily addresses 3 major use cases in daily operations, for example:
1. The super admin has the responsibility/permission to maintain the "available versions of the Control Plane" to be used by tenant admins -- btw, we do allow the tenant admins to spin up new tenant masters, don't we?
2. The tenant admin spins up tenant clusters by simply referring to any of the "available versions of the Control Plane" -- just like claiming PVC against a predefined PV, I guess.
3. The super admin has the responsibility/permission to deprecate older/vulnerable versions to avoid too much flexibility granted to tenant admins if they can literarily define any K8s version to build the tenant masters -- this sounds like giving a way to tenant admins to hack into the super cluster -- at least it's not a good idea to maintain the manageability. What do you guys think?

> type NestedControlPlaneTemplateSpec struct {

Author: @brightzheng100 I agree that `template` can be helpful to manage different settings. But in the first version, we may just implement something more generic/flexible, i.e., creating each CR independently, and we can build the `template` mechanism on top of it later.

> type NestedControlPlaneTemplateSpec struct {

Reviewer 2: Is the template concept going to be implemented by CAPN itself? Or is this something that we're expecting upstream Cluster API to implement?

> type NestedControlPlaneTemplateSpec struct {

Reviewer 3: Agree @brightzheng100 and @charleszheng44. These are all functions that can layer later to make the DX improved but would be hard to treat as a two-way door if we forced that style first.

@vincepri I believe that will still end up in CAPN as a way to augment and make CAPN easier to consume for some users  like Bright has brought up, which is where VC makes the DX really simple but forces all implementations to implement copies and makes upgrades somewhat complex because there isn't a single owner of each resource to maintain updating of that component.

    // other fields ...    
    

>     // NestedEtcd contains information required by the etcd provider to create the Etcd
>     NestedEtcd NestedComponentSpec `json:"nestedEtcd"`
> 
>     // NestedKAS contains information required by the apiserver provider to create the apiserver
>     NestedKAS NestedComponentSpec `json:"nestedKAS"`
> 
>     // NestedKCM contains information required by the controller-manager provider to create the controller-manager
>     NestedKCM NestedComponentSpec `json:"nestedKCM"`

Reviewer 2: ```suggestion
    // Etcd contains information required by the etcd provider to create the Etcd
    Etcd NestedComponentSpec `json:"etcd"`

    // ApiServer contains information required by the apiserver provider to create the apiserver
    ApiServer NestedComponentSpec `json:"apiServer"`

    // ControllerManager contains information required by the controller-manager provider to create the controller-manager
    ControllerManager NestedComponentSpec `json:"controllerManager"`
```

Given that all of these are already in a `Nested` struct

}
```

> We will let users specify component providers through NCP, implement a native mode that has each sub-controller to create the component on the super cluster, and define three new types `EtcdProvider`, `KASProvider`, and the `KCMProvider`:

Reviewer 2: Could we run all controllers all the time instead of allowing to turn them on/off? If controllers aren't reconciling anything because there are no resources / events, it should be fine?

Reviewer 3: The idea here would be that in the config (or flags 🤮 ) we can turn off specific controllers and if you wanted to use an external etcd cluster provider like `etcdadm` or `etcd-cluster-operator` you can choose to flip off the built-in provider and you just supply your own controller listening for `NEtcd` and creating w/e the necessary components for that implementation. This will change a bit of the implementation as per - https://github.com/kubernetes-sigs/cluster-api-provider-nested/pull/11/files/4b6d8aea98ed04eea543f2e29319567d4bab5207#r521166698

```go

> type KASProvider string

Author: Is it possible that for a single component, there exist different kinds of component contollers on the super cluster at the same time? And different users may wanna use different controllers?

> type KASProvider string

Reviewer 3: I think that is a really novel use case but I don't think it's something we have to build into the first implementation. Removing upfront doesn't preclude us from adding this later as those use cases develop.

> type KASProvider string

Author: In the first implementation, let's assume that there will be only one component controller for each component at a time.

type EtcdProvider string

type KCMProvider string

const (
    NativeKASProvider KASProvider = "NativeKASProvider"

    NativeEtcdProvider  EtcdProvider = "NativeEtcdProvider"
    EtcdClusterOperator EtcdProvider = "EtcdClusterOperator"
    EtcdOperator        EtcdProvider = "EtcdOperator"

    NativeKCMProvider KCMProvider = "NativeKCMProvider"
)
```

We will add three new fields `NestedEtcdProvider`, `NestedKASProvider`, and the `NestedKCMProvider` into the NCP, which will make the NCP look like the following:

```go
// NestedControlPlaneSpec defines the desired state of NestedControlPlane

> type NestedControlPlaneSpec struct {
>     // other fields.... 
>     
>     // NestedEtcdProvider specifies which EtcdProvider will be used to create the Etcd
>     // +optional
>     NestedEtcdProvider EtcdProvider `json:"etcdProvider,omitempty"`
>      
>     // NestedKASProvider specifies which KASProvider will be used to create the kube-apiserver
>     // +optional
>     NestedKASProvider KASProvider `json:"kasProvider,omitempty"`
>  
>     // NestedKCMProvider specifies which KCMProvider will be used to create the kube-controller-manager
>     // +optional
>     NestedKCMProvider KCMProvider `json:"kcmProvider,omitempty"`
> }

Reviewer 2: ```suggestion
type NestedControlPlaneSpec struct {
    // other fields.... 
    
    // EtcdProvider specifies which EtcdProvider will be used to create the Etcd
    // +optional
    EtcdProvider EtcdProvider `json:"etcdProvider,omitempty"`
     
    // ApiServerProvider specifies which KASProvider will be used to create the kube-apiserver
    // +optional
    ApiServerProvider KASProvider `json:"apiServerProvider,omitempty"`
 
    // ControllerManagerProvider specifies which KCMProvider will be used to create the kube-controller-manager
    // +optional
    ControllerManagerProvider KCMProvider `json:"controllerManagerProvider,omitempty"`
}
```

Same suggestion as above regarding naming.

On a more general note, do we need pluggable providers for the first iteration? Would it be ok to reduce scope a bit and only allow the built-in controllers for etcd/api-server/controller-manager instead?

Reviewer 3: We don't fully as it's documented IMO, this was called out a couple places. Specifically here - https://github.com/kubernetes-sigs/cluster-api-provider-nested/pull/11/files/4b6d8aea98ed04eea543f2e29319567d4bab5207#r521166698 As long as we have a way to swap the controllers for specific implementations that would be sufficient, IE I could use an etcd prefix or an operator backing the `NEtcd` controller instead and manage that out-of-tree.

I'm very hesitant to add too much pluggability upfront except controlling what controller manages each object. Since there are so much prior art in the OSS for etcd controllers and we should be able to support those without forking.

```

### Bootstrap

#### Create prerequisites 

The NCP controller will need to create some resources before sub-controllers start creating the real NCP components.

> * **ClusterNamespace** is used to store all resources for the NCP.

Reviewer 3: From a Namespacing perspective what do you think about Namespaces being done manually, meaning if you are deploying an NCP into a cluster you create the namespace and create the CRs in that namespace meaning we don't need to deal with Namespace auth delegation from each one of the controllers. Then when we come back around to templating we can add back in the automated Namespace creation into the NCPT, would that work? What do you think?

Reviewer 4: I think in this proposal, we don't need to rely on NCP controller to create all CRs and namespace. We can just assume they are there when controller work. We can leave the discussion to NCP design.

Reviewer 3: Agree, that makes the most sense to me. We'll need to update line 187 to reflect that. 💯

> * **NKAS service**  is used to visit the NKAS. Since the IP/domain name is required when generating TLS certificates and kubeconfigs of NKAS. This service will need to be created in advance.

Author: Yes. let's make the KAS controller to create the service.

> * **NestedCRs**, including NEtcd, NKAS, and NKCM. are CRs that hold information required by the sub-controller. 

Reviewer 3: Let's start without NCP creating the nested CRs since we won't have templates, so to create a nested cluster you need to create each of NCluster, NCP, NEtcd, NKAS, NKCM and these will work together to create the control plane.

After creating the above prerequisites, the sub-controller will start creating the workload of each component.

> As there exist dependencies between components, i.e., KAS cannot run without Etcd, KCM cannot work without KAS, when creating the NCP component, sub-controllers will need to get information and status of other CRs. Thereby, we will have each nested component CR points to the NCP CR. Moreover, To get other CRs through the NCP CR, we suggest adding object references of the three Nested CRs into the `NCP.status`, which will make the `NCP.status` look something like the following:

Reviewer 3: If we keep CRs and NCP's in the same namespace we could use actual `metadata.ownerReferences`

Author: Are we going to assume that CRs and NCP will be located in the same namespace? As we intend to not create the `clusternamespace` for the user, should we introduce a new mechanism to ensures that the CRs and the NCP will be located in a same namespace?

Reviewer 3: Yeah, `OwnerReferences` themselves will be forced to only work within the same namespace. (you can't associate to another NS). I think at this foundational level we should just expect that all CRs for a control plane `netcd, nas, ncm, ncp and nc (nested cluster)` are in the same namespace as where the components are provisioned.

Do you think that would cause any issues?

Author: No. I just worried that users may accidently put CRs in different ns, for example, forgetting to specify the ns and result in the CR being located in the default ns.

```go
// NestedControlPlaneSpec defines the desired state of NestedControlPlane
type NestedControlPlaneStatus struct {
    // other fields.... 
    
    // NestedEtcd is the reference of the NestedEtcd CR which is under the charge of this NestedControlPlane
    // +optional
    NestedEtcd *corev1.ObjectReference `json:"nestedEtcd,omitempty"`
     
    // NestedKAS is the reference of the NestedKAS CR which is under the charge of this NestedControlPlane
    // +optional
    NestedKASP  *corev1.ObjectReference `json:"nestedKAS,omitempty"`
 
    // NestedKCM is the reference of the NestedKCM CR which is under the charge of this NestedControlPlane
    // +optional
    NestedKCM *corev1.ObjectReference `json:"nestedKCM,omitempty"`
}
```

### Creation

End-users can specify the component provider in the NCP, and the NCP controller will add the provider's information to the corresponding Nested CR. 

#### Native way 

> If no component provider is specified, the sub-controller will create the component under the native mode, which will create the component using the StatefulSet workload. The readiness and liveness probe will be used, and we will mark each component as ready only when the corresponding Statefuleset is ready. As the KAS cannot work without available Etcd and the KCM cannot run without KAS, the three components need to be created in the order of NEtcd, NKAS, and NKCM

Author: Will describe the alternative in the proposal.

We will host sets of default templates in this repository. Users can specify which set of templates they intend to use by specifying the corresponding version of the `CommonSpec` in the NCPT.

The creating process will include six steps:
![Control Plane Creating Process](nativeway.png)

1. The user creates the NCP CR. 

> 2. NCP controller creates ClusterNamespace, NEtcd, NKAS, NKCM, and the NKAS service.

Author: Will let the KAS controller create the KAS service

> 3. NEtcd controller creates the NEtcd StatefulSet, and a headless service (as the NKAS will only visit the NEtcd from inside the cluster) that exposes the NEtcd. Once the NEtcd StatefulSet is ready, the NEtcd controller will mark the NEtcd as ready and store the Etcd's certificates into the PKI configmap.

Author: > CAPI cert resources are documented https://cluster-api.sigs.k8s.io/tasks/certs/using-custom-certificates.html for more information
That's great! then I think we don't need the PKI configmap anymore.

> 4. NKAS controller creates the TLS certificate, kubeconfig required by KCM, an admin-kubeconfig for end-users, and the NKAS workload. Same as above, we mark the NKAS as ready once the NKAS StatefulSet is ready, store the certificates, and the kubeconfig into the PKI configmap.

Author: Yes, as we can access the KAS CA and the KAS service, we can let the KCM controller create the kubeconfig.

> 4. NKAS controller creates the TLS certificate, kubeconfig required by KCM, an admin-kubeconfig for end-users, and the NKAS workload. Same as above, we mark the NKAS as ready once the NKAS StatefulSet is ready, store the certificates, and the kubeconfig into the PKI configmap.

Reviewer 3: Awesome :+1:

5. NKCM controller creates the NKCM StatefulSet, and marks the NKCM as ready once the StatefulSet is ready.

> 6. After all the three components are ready, the NCP controller marks the NCP CR as ready.

Author: Is `ControlPlaneEndpoint` the network endpoint that will be used to access the NCP?

> 6. After all the three components are ready, the NCP controller marks the NCP CR as ready.

Reviewer 3: Yes, check out https://cluster-api.sigs.k8s.io/developer/providers/cluster-infrastructure.html?highlight=ControlPlaneEndpoint#data-types for more information.

#### Using the third-party provider

If users intend to use a third-party component provider, extra codes need to be added to the sub-controller. Most third-party providers require users to create a CR to let the provider start working. For example, Etcd-cluster-operator requires the [EtcdCluster](https://github.com/improbable-eng/etcd-cluster-operator/blob/master/api/v1alpha1/etcdcluster_types.go). Therefore, we added a field, `ProviderObject`, into each Nested CR's status. This object reference will point to the object/CR that is required by the designated provider.

In the following example, we use the Etcd-cluster-operator as the Etcd provider to create the Etcd. The creating process will include eight steps:
![Creating a Control Plane using third-party provider](third-party-provider.png)

1. NCP controller perceives the creation of a new NCP CR.

> 2. NCP controller refers to the corresponding NCPT and creates Nested CRs.

Reviewer 3: Update this without NCPT

3. NEtcd controller creates an EtcdCluster CR.

4. The EtcdClusterOperator created the Etcd workload based on the EtcdCluster CR and updated the EtcdCluster as stable/ready.

5. NEtcd controller notices that the EtcdCluster is changed to stable/ready, mark the NEtcd as ready, and stores the NEtcd's PKI into the PKI Configmap.

6. NKAS controller creates the workload of NKAS, updates the NKAS and PKI Configmap.

7. NKCM controller creates the workload of the NKCM, update NKCM and PKI Configmap.

8. Once all the three components are ready, the NCP controller marks the NCP CR as ready.

### Control Plane Custom Resources 

The followings are CRDs of the three components.

### NestedEtcd CRD

```go
// NestedEtcdSpec defines the desired state of NestedEtcd
type NestedEtcdSpec struct {
    // NestedControlPlane is the control plane that owns the NestedEtcd 
    // +optional

>     NestedControlPlane *corev1.ObjectReference `json:nestedControlPlane,omitempty`

Reviewer 3: ```suggestion
    ControlPlaneRef *corev1.ObjectReference `json:nestedControlPlane,omitempty`
```

    
    // NestedComponentSpec contains the common and user-specified information that are 
    // required for creating the component
    // +optional
    NestedComponentSpec NestedComponentSpec  `json:nestedComponentSpec,omitempty`

    // Provider defines which provider will be used to create the etcd
    // +optional

>     Provider EtcdProvider `json:provider,omitempty`

Author: will remove this field

    Ready bool `json:"ready,omitempty"`

    // ProviderObject is the object that used by the underlying provider.
    // Under the native mode, the ProviderObject points to the StatefulSet
    // that contains the component's pods.
    ProviderObject *corev1.ObjectReference `json:referralObject,omitempty`

    // EtcdDomain defines how to address the etcd instance
    Addresses []NestedEtcdAddresses `json:"addresses,omitempty"`
}

// NestedEtcdAddresses defines the observed addresses for etcd
type NestedEtcdAddresses struct {
    // IP Address of the etcd instance.
    // +optional
    IP string `json:"ip,omitempty"`
    
    // Hostname of the etcd instance
    Hostname string `json:"hostname,omitempty"`
    
    // Port of the etcd instance
    // +optional
    Port int32 `json:"port"`
}

// NestedEtcd is the Schema for the nestedetcds API
type NestedEtcd struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`
    
    Spec   NestedEtcdSpec   `json:"spec,omitempty"`
    Status NestedEtcdStatus `json:"status,omitempty"`
}
```

### NestedKAS CRD

```go
type NestedKASSpec struct {
    // The NestedControlPlane that owns this NestedKAS

>     NestedControlPlane *corev1.ObjectReference `json:nestedControlPlane,omitempty`

Reviewer 3: ```suggestion
    ControlPlaneRef *corev1.ObjectReference `json:nestedControlPlane,omitempty`
```

   
    // NestedComponentSpec contains the common and user-specified information that are 
    // required for creating the component
    // +optional
    NestedComponentSpec NestedComponentSpec  `json:nestedComponentSpec,omitempty`

    // Provider defines which provider will be used to create the kube-apiserver 
    // +optional
    Provider KASProvider `json:provider,omitempty`
}

// NestedKASStatus defines the observed state of NestedKAS
type NestedKASStatus struct {
    // Ready is set if all resources have been created
    // +kubebuilder:default=false
    Ready bool `json:"ready,omitempty"`

    // ProviderObject is the object that used by the underlying provider.
    // Under the native mode, the ProviderObject points to the StatefulSet
    // that contains the component's pods.

>     ProviderObject *corev1.ObjectReference `json:referralObject,omitempty`

Author: I think we may have to keep this reference. We can set the `metav1.OwnerReference` of the provider object as the component CR, but the component CR also need to get the provider object.

>     ProviderObject *corev1.ObjectReference `json:referralObject,omitempty`

Reviewer 3: That is a good point, maybe this should be changed to:

```suggestion
    ProviderRef *corev1.ObjectReference `json:"providerRef,omitempty"`
```

    // EtcdDomain defines how to address the etcd instance
    // +optional

>     KASService *corev1.Service `json:"addresses,omitempty"`

Author: will change to the objectreference

}

// NestedKAS is the Schema for the NestedKASs API
type NestedKAS struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`
    
    Spec   NestedKASSpec   `json:"spec,omitempty"`
    Status NestedKASStatus `json:"status,omitempty"`
}
```

### NestedKCM CRD

```go
// NestedKCMSpec defines the desired state of NestedKCM
type NestedKCMSpec struct {
    // The NestedControlPlane that owns this NestedKCM

>     NestedControlPlane *corev1.ObjectReference `json:nestedControlPlane,omitempty`

Reviewer 3: ```suggestion
    ControlPlaneRef *corev1.ObjectReference `json:nestedControlPlane,omitempty`
```

Author: Will remove the ControlPlaneRef, as we are going to use the OwnerReference

    
    // NestedComponentSpec contains the common and user-specified information that are 
    // required for creating the component
    // +optional
    NestedComponentSpec NestedComponentSpec  `json:nestedComponentSpec,omitempty`

    // Provider defines which provider will be used to create the kube-controller-manager
    // +optional
    Provider KCMProvider `json:provider,omitempty`
}

// NestedKCMStatus defines the observed state of NestedKCM
type NestedKCMStatus struct {
    // Ready is set if all resources have been created
    Ready bool `json:"ready,omitempty"`

    // ProviderObject is the object that used by the underlying provider.
    // Under the native mode, the ProviderObject points to the StatefulSet
    // that contains the component's pods.
    ProviderObject *corev1.ObjectReference `json:referralObject,omitempty`
}

// NestedKCM is the Schema for the NestedKCMs API
type NestedKCM struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`
    
    Spec   NestedKCMSpec   `json:"spec,omitempty"`
    Status NestedKCMStatus `json:"status,omitempty"`
}
```

### Security Model

> Creating an NCP requires the end-user to submit a creation request, and the 

Author: will add the definition in the glossary.

cluster administrator will be responsible for creating the NCP CRs and applying 
them. Once the NCP is ready, the cluster administrator will return a kubeconfig 
to the end-user, as each end-user can only access the apiserver assigned to he/she. 
There is no need to worry about malicious users to manipulate other users' resources. 
A malicious user can still skew the system by creating a massive amount of resources. 
To avoid this, we need to enhance the syncer component; however, this topic is 
beyond this proposal's scope. The proposed mechanism will not lead to any 
severe security issue.

## Alternatives

Users may implement their own NCP controller that creates master components 
through a third-party platform. For example, users can have an on-premise super 

> cluster and a bunch of NCP hosted on the public computing platform.

Author: Will remove this section.
